src/op_mode/qos.py

```diff
@@ -0,0 +1,227 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# This script parses output from the 'tc' command and provides table or list output
+
+import sys
+import typing
+import json
+from tabulate import tabulate
+
+import vyos.opmode
+from vyos.configquery import ConfigTreeQuery
+from vyos.utils.process import cmd
+
+conf = ConfigTreeQuery()
+
+def detailed_output(dataset, headers):
+    for data in dataset:
+        adjusted_rule = data + [""] * (len(headers) - len(data)) # account for different header length, like default-action
+        transformed_rule = [[header, adjusted_rule[i]] for i, header in enumerate(headers) if i < len(adjusted_rule)] # create key-pair list from headers and rules lists; wrap at 100 char
+
+        print(tabulate(transformed_rule, tablefmt="presto"))
+        print()
+
+def get_tc_info(interface_dict, interface_name, policy_type):
+    policy_name = interface_dict.get(interface_name, {}).get('egress')
+    class_dict = conf.get_config_dict(['qos', 'policy', policy_type, policy_name], key_mangling=('-', '_'),
+                            get_first_key=True)
+
+    return policy_name, class_dict
+
+def format_data_type(num, suffix):
+    if num < 10**3:
+        return f"{num}  {suffix}"
+    elif num < 10**6:
+        return f"{num / 10**3:.3f} K{suffix}"
+    elif num < 10**9:
+        return f"{num / 10**6:.3f} M{suffix}"
+    elif num < 10**12:
+        return f"{num / 10**9:.3f} G{suffix}"
+    elif num < 10**15:
+        return f"{num / 10**12:.3f} T{suffix}"
+    elif num < 10**18:
+        return f"{num / 10**15:.3f} P{suffix}"
+    else:
+        return f"{num / 10**18:.3f} E{suffix}"
+
+def show_shaping(raw: bool, ifname: typing.Optional[str], classn: typing.Optional[str], detail: bool):
+    # Scope which interfaces will output data
+    if ifname:
+        interface_dict = {}
+        interface_dict[ifname] = conf.get_config_dict(['qos', 'interface', ifname], key_mangling=('-', '_'),
+                            get_first_key=True)
+    else:
+        interface_dict = conf.get_config_dict(['qos', 'interface'], key_mangling=('-', '_'),
+                            get_first_key=True)
+
+    raw_dict = {'qos': {}}
+    for i in interface_dict.keys():
+        interface_name = i
+        output_list = []
+        output_dict = {'classes': {}}
+        raw_dict['qos'][interface_name] = {}
+
+        # Get configuration node data
+        policy_name, class_dict = get_tc_info(interface_dict, interface_name, 'shaper')
+
+        class_data = json.loads(cmd(f"sudo tc -j -s class show dev {i}"))
```

Comment:
Is the standard method using the vyos.ifconfig.Interface.exists function? Or just throw that block in a try/except?

Reply:
I think the completion help for the XML will be enough, for example  https://github.com/vyos/vyos-1x/blob/03d8c16a12d182a2fb2ab9f7acec5d8cb83937ba/op-mode-definitions/lldp.xml.in#L19-L21

Reply:
I already have the <path> completionHelper, since there's no need to list interfaces that QoS isn't configured on, but the completionHelper won't prevent the user from typing an interface that isn't present. I'd love if the op-mode-definitions had a <constraint> element as well.

Reply:
Done, check added!